spec/praxis/request_stage_spec.rb

```diff
@@ -1,93 +1,181 @@
+        stage.execute
+
+        expect(controller.response).to be(response)
       end
+
     end
   end
 
-  context ".execute" do
-    let(:double_stage_ok){ double("ok stage") }
+  context "#run" do
 
-    context 'when all stages suceed' do
+    let(:before_callbacks) { double('before_callbacks') }
+    let(:after_callbacks) { double('after_callbacks') }
+
+    let(:controller_before_callbacks) { double('controller_before_callbacks') }
+    let(:controller_after_callbacks) { double('controller_after_callbacks') }
+
+    after do
+      stage.run
+    end
+
+    context 'callback execution' do
       before do
-        expect(double_stage_ok).to receive(:run).twice
-        stage.instance_variable_set(:@stages, [double_stage_ok, double_stage_ok])
-      end
-      
-      it "runs them all" do
-        stage.execute
+        allow(stage).to receive(:before_callbacks).once.and_return(before_callbacks)
+        allow(stage).to receive(:after_callbacks).once.and_return(after_callbacks)
+
+        allow(controller_class).to receive(:before_callbacks).once.and_return(controller_before_callbacks)
+        allow(controller_class).to receive(:after_callbacks).once.and_return(controller_after_callbacks)
       end
-      it 'returns nil' do
-        expect( stage.execute ).to be(nil)
+
+      it "sets up and executes callbacks" do
+        expect(stage).to receive(:setup!)
+        expect(stage).to receive(:setup_deferred_callbacks!)
+        expect(stage).to receive(:execute)
+        expect(stage).to receive(:execute_callbacks).once.with(before_callbacks)
+        expect(stage).to receive(:execute_callbacks).once.with(after_callbacks)
+        expect(stage).to receive(:execute_controller_callbacks).once.with(controller_before_callbacks)
+        expect(stage).to receive(:execute_controller_callbacks).once.with(controller_after_callbacks)
+        #stage.run
       end
+
     end
-    context 'when one stage returns a Response' do
-      it "runs no further stages after that" do
-        double_stage_fail = double("fail stage")
-        expect(double_stage_ok).to receive(:run).once
-        expect(double_stage_fail).to receive(:run).once.and_return(Praxis::Responses::Ok.new)
-        stage.instance_variable_set(:@stages, [double_stage_ok, double_stage_fail, double_stage_ok])
-        stage.execute
+
+    context 'when the before execute_controller_callbacks return a Response' do
+      let(:action_name) { :show }
+      let(:response) { Praxis::Responses::Unauthorized.new }
+
+      before do
+        expect(stage).to receive(:execute_controller_callbacks).once.and_return(response)
       end
+
+
+      it 'does not call "execute"' do
+        expect(stage).to_not receive(:execute)
+      end
+
+      it 'does not execute any "after" callbacks' do
+        expect(stage).to_not receive(:after_callbacks)
+        expect(controller_class).to_not receive(:after_callbacks)
+      end
+
+      #it 'assigns controller.response' do
```

Comment:
Right, the value's ultimately set, just that it's set in execute_controller_callback, so doesn't make sense in the "#run" context is all.